library/src/scala/IArray.scala

```diff
@@ -250,8 +250,8 @@ object IArray:
     genericArrayOps(arr).takeWhile(p)
 
   /** Returns a mutable copy of this immutable array. */
-  extension [T](arr: IArray[T]) def toArray: Array[T] =
-    arr.clone.asInstanceOf[Array[T]]
+  extension [T](arr: IArray[T]) def toArray: Array[? <: T] =
+    arr.clone
```

Comment:
This is a source-breaking change, so it needs to go in a minor release. (but it won't be detected by MiMa)

While this fixes the unsoundness, it makes the API barely usable as a result. The point is to return a mutable copy, i.e., something that the user can mutate. But if you give them an `Array[? <: T]`, they can't actually mutate it. IMO a correct fix would be to take a `using ClassTag[T]` to be able to return an actual `Array[T]`. Of course that's a more complicated fix, since we'll need to introduce a hidden overload that's there only for binary compatibility.

Reply:
Type inference changes are also usually source-breaking in some way but we don't bump to minor for them. Anyway we can also just deprecate that method and introduce a working overload later

Reply:
> Type inference changes are also usually source-breaking in some way but we don't bump to minor for them.

We should. Or we shouldn't advertise that minor versions preserve source compatibility.

Anyway, this case is particularly egregious, since it *intentionally* breaks source compatibility. If that doesn't trigger a minor version bump, then we *must* stop saying that minor versions preserve source compatibility.

> Anyway we can also just deprecate that method and introduce a working overload later

It's unclear that we can introduce a public overload taking a `using` parameter. It wouldn't take precedence over this one which takes no evidence. :-(

Reply:
>  Or we shouldn't advertise that minor versions preserve source compatibility.

I don't think we should since guaranteeing that is incredibly hard (scala 2 doesn't either).

> It wouldn't take precedence over this one which takes no evidence. :-(

Indeed, they're ambiguous, but it turns out things still work out because there's an implicit conversion from IArray to ArraySeq, see updated commit.